### ISSUE TITLE
Love the idea

### ISSUE THREAD

**Issue by pmwoodward3** [jrudio/one-time-plex#1]
Love this idea! I totally run into this and need something like this to help. 

Only idea I would suggest is some sort of "token" system, where I could give someone X amount of "view tokens" they can use per day/week. That would really help put my mind at ease when giving access to my plex server who are friends of friends.

Keep up the great work!

**Comment by jrudio** [comment]
Great suggestion! I definitely want to integrate a token system.

I first need to work on a simple front-end to manage: users, users' requests, app settings, then, I think I should be able to work on the token system.

**Comment by pmwoodward3** [comment]
Awesome! Thanks for the quick reply. Think you could cannibalize some of those systems from plexpy or plex requests to get a jump start? I know plexpy just added the ability for you to designate certain users to view/user permissions.

**Comment by jrudio** [comment]
Yeah, I can take some pointers from their projects to expedite this a little bit. 

Currently, I have that implemented, but requires the friend-of-a-friend to have a plex account. In the readme it's noted to make a second account, but should be more apparent. 

Also, it currently requires a Plex Pass (for the pms owner), which I think I can eventually work around, but would require the friend-of-friend to give a little more info (possibly ip, machine id). The reason for the plex pass requirement is that I restrict media via labels, which helps users not view anything else in your library.

I checked out plexpy's changelog and it says they use the guest feature. I remember trying that out before deciding to just create another plex account. I forgot the reason for deciding against it, but I'll revisit it and see what I can do.

**Comment by pmwoodward3** [comment]
Hey, looks like you're just powering through commits!!! Just wanted to check in and see how far along you feel you are at the moment, just want to know when it's far enough along that I could spin this up and help out!

Thanks again,
Pat

**Comment by jrudio** [comment]
Hi Pat, thanks!

I'm pretty far from where I would say it is usable. Admittedly, I have been using this project to learn redux so I have been going much slower than anticipated, but it's looking good so far.

What I would consider usable is to implement these:

- Search and grab media info from server; for the rating key, so when the user is monitored it can be checked against the current sessions.

  - For the most part this is straightforward to implement

- Authentication and authorization for users

- Monitoring multiple users

As far as an estimate goes, maybe a week? I honestly have no clue.

I have quite a few ideas to implement that may not be necessary, but would like a different perspective. What do you think is needed for a basic setup? How do you envision the process would be from your end?

**Comment by pmwoodward3** [comment]
Sorry for the delay, life has been crazy.

I'm not a programmer myself, but i'm a sysadmin who knows enough to break things haha.

I found this the other day and i'm currently using it, it seems to do a pretty good job at pulling all the metadata from the plex items as well as authentication. Might be worth it to check out how they do it.

https://forums.plex.tv/discussion/218740/beta-release-php-library-presenter-for-plex-plpp-v0-8-beta/p1

Also, just thinking outloud. What if instead of making people have plex accounts to use "One Time Plex", the server owner created a OTP account that the system used, if that makes any sense. That way you only have one account talking to the server at anytime and it would be really easy to revoke that user's permissions quickly if something happened. This might not be the most efficient way to do it, I have no idea.

Also, if you want to bounce any ideas off me just let me know, I have a pretty big plex server (12tb) and ~40 friends using it, so I would think i'm a pretty good use case.


Anyway, really excited still to see this in action! :)

**Comment by jrudio** [comment]
Sweet, I checked out that project and it has a nice ui to it. I think I might go with a simple grid to browse media (can make this optional or private), and then also a search bar to quickly search via PMS for certain media.

For the most part I know how to grab the media and such as long as you (the one time plex admin) either supplies the plex token (if required; your one time plex and PMS might be on the same network and may not need a token) or plex credentials to fetch one. The thing is I wanted to utilize Plex's https that comes with our servers, but that would make a requirement of the plex token. Not sure if that would be bad.

>What if instead of making people have plex accounts to use "One Time Plex", the server owner created a OTP account that the system used, if that makes any sense. That way you only have one account talking to the server at anytime and it would be really easy to revoke that user's permissions quickly if something happened.

Sorry, I'm slightly confused. 😀  

I make mention of having the server owner creating a `<username>-guest` account as a pre-requisite, which I think is what you mean by OTP using an account the system used (please correct me if i'm wrong, lol). Having an end user use their own account would make monitoring pretty simple, which would just be comparing pms' current sessions with the friend id and media id, to what was requested.

If we were to use just one account for all users, we would need to make note of their ip address to check against, as that is what `GET /status/sessions` returns, to differentiate each user and the media they requested. It may also return the machine-id, player type, and such.

Sorry, for progressing so slow the backend is being written in go and there isn't really anything like Django or MeteorJS to get a simple auth system going. Where I'm at right now I think I'm going to jump ahead to work on the automatic setup of pms. The idea is you (as an admin) just enter your plex login or plex token, then it will fetch the https url to your pms and will make all PMS calls via that url.

A simple roadmap after that would be:

- Work on concurrent monitoring (So far I've only tested with one user)
- Automate media restriction via labels (Already have this setup via my [go-plex-client](https://github.com/jrudio/go-plex-client) I just need to implement into otp)
- Manage user requests
- And maybe something minor I'm missing


>Also, if you want to bounce any ideas off me just let me know,

Will do!

> I have a pretty big plex server (12tb) and ~40 friends using it, so I would think i'm a pretty good use case.

Got Damn. I'm running an r710 that gets barely any love from my family. What kind of hardware do you have? How many simultaneous users?

**Comment by jrudio** [comment]
Whoops, didn't realize I wrote such a huge wall of text. Sorry!

**Comment by pmwoodward3** [comment]
Haha no problem, as far as the server goes, mine is a HP with two xeon L5639 (6 core, 12 thread each) with ~32gb of ram. Honestly I think the magic is to have all your media in a format that will minimize the amount of transcoding the server needs to do, along with a gigabit connection for my server which just gives me way more headroom for sending media to people, if that makes any sense. 

I've also been running my server for 3-4 years now and have moved it from Windows to Linux and rebuilt it several times, so it's in a pretty great sweet spot now as far as performance.

**Comment by jrudio** [comment]
@pmwoodward3 Sorry for the lack of updates. I'm currently working on the monitoring portion to support multiple users of the same plex account. I'm curious as to how you have your users setup? Do they each have their own plex account? Or do they access your server via Plex home under one Plex account?

Also, if you want to fire this up just know that the monitoring portion is still being worked on.

**Comment by pmwoodward3** [comment]
Currently all my users have their own account, just like the default setup.

And awesome! Congrats on the progress :D 

I'm trying to set OPT up right now and i'm not seeing ./opt like you have in the readme, is there something I need to do first after I clone it locally?

**Comment by jrudio** [comment]
Whoops, I forgot to mention in the README that it needs to be compiled with
Go. I can build for all platforms and upload to the Github release section.
If you don't want to install Go.

Here's the Go website https://golang.org/dl/

**Comment by pmwoodward3** [comment]
I have go installed on ubuntu x64, if you have the commands though I could compile it myself :)

**Comment by pmwoodward3** [comment]
# go build
# _/opt/one-time-plex
./main.go:33: err.Error undefined (type chan string has no field or method Error)

is what I get currently

**Comment by jrudio** [comment]
Damn it I didn't commit the updated version.

**Comment by jrudio** [comment]
Ok pushed main.go, and also included some instructions and current limitations of otp in the readme.
